gkp/gkp_tf_env/gkp_tf_env.py

```diff
@@ -483,14 +471,19 @@ def batch_size(self, size):
             self._batch_size = size
         except:
             raise ValueError('Batch size should be positive integer.')
-    
-    @property 
+
+    @property
+    @abstractmethod
     def N(self):
-        return self._N
-    
-    @N.setter
-    def N(self, n):
-        if 'code_map' in self.__dir__():
-            raise ValueError('Cannot change N after initialization.')
-        else:
-            self._N = n
+        """
+        Size of the oscillator Hilbert space truncation (not including the qubit)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def tensorstate(self):
```

Comment:
move `tensorstate` to `gkp_init` ?

Reply:
I left this here since this allows the class itself to encapsulate `tensorstate`, so that gkp_init doesn't need to know.

It could be a variable set on init but the property + abstractmethod seems like an acceptable way to enforce the interface contract more explicitly.